### PR TITLE
Do video models dream of electrical sheeps in motion

### PR DESCRIPTION
**Pull request by daniel-j-h**
Work in progress; let's see if this works out :hugs:

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Here's where we are right now:
1. Create a random video with 32 frames of shape CxTxHxW 3x32x112x112
2. Hook this random tensor up in the computational tree so we can optimize it
3. Pass this random tensor through the trained video model up to the nth layer
4. Create a loss maximizing the the nth layer's activations by optimizing the input video

Here are example videos maximizing layer3 and layer4 activations:

ex02-layer3-upscale

ex01-layer4-highlr-upscale

Learnings
- Can use a very high learning rate e.g. `1e-2`
- Should maximize activations for a single channel / volume, not mean over all
- Get high-frequency patterns fooling the network but not reasonable for humans

We then looked into regularization terms to get rid of the high-frequency patterns; the total variation regularization loss implemented right now results in very strong checkerboard patterns.

ex02-layer3-regscale-upscale

ex03-layer1-noreg-up

Next actions: figure out better regularization term and look into prior art how other folks handle this.

**Comment by daniel-j-h** [comment]
Completely different approach now; below are results for maximizing layer2 activations with different learning rates and number of iterations (second one is stronger) - I think this is the way to go :rocket:


dream-y

dream-final


<details>

<summary>More examples from: stem, layer1, layer2, layer3, layer4</summary>

dream-0

dream-1

dream-2

dream-3

dream-4


---

dream-00

dream-01

dream-02

dream-03

dream-04

</details>

**Comment by sandhawalia** [comment]
Dayum 🔥

**Comment by daniel-j-h** [comment]
Here are a few examples when we - instead of optimizing all activations - select a specific channel; the activations are of shape NxCxTxHxW; below are a few examples for optimizing with fixed `i` in

    acts[:,i, :, :, :].norm()

dream-17-up
dream-16-up
dream-15-up
dream-14-up
dream-13-up
dream-12-up
dream-11-up
dream-10-up

**Comment by daniel-j-h** [comment]
Starting with a random tensor instead of a seed video and optimizing all activations:

dream-rnd-up

**Comment by daniel-j-h** [comment]
Starting from random tensor optimizing specific channels (3 and 6 in this case)

dream-zz-up
dream-z-up

**Comment by daniel-j-h** [comment]
Bringing back the 3d total variation loss term and scaling it

dream-tvn3-up

seems to get rid of high frequencies in the output which makes it more pleasant to look at :hugs: 

direct comparison:
- [no total variation term](https://user-images.githubusercontent.com/527241/66868637-fe7cb380-ef9d-11e9-93d5-d46ebbf15f00.gif)
- [(-1) * 100 * tv(video)](https://user-images.githubusercontent.com/527241/66871558-2e2eba00-efa4-11e9-8f11-0bed6ed0152f.gif)
-  [(-1) * 200 * tv(video)](https://user-images.githubusercontent.com/527241/66871922-00964080-efa5-11e9-9a7e-ec9b583785ec.gif)

**Comment by daniel-j-h** [comment]
Changing how we normalize the gradients

primer1-up

**Comment by daniel-j-h** [comment]
Latest version maximizing all channels in layer2

primer1-up

Latest version maximizing channel 6 in layer2

primer1-up

**Comment by daniel-j-h** [comment]
Here are results for
- layer2 again
- maximizing specific TxHxW channel activations
- resizing to frame size of 248 px on the shorter edge

which barely fits on one of my GTX 1080 TIs:

royal-wedding

royal-wedding

royal-wedding


Input clip from

    youtube-dl -f 18 yJbXdOdTaJc

    ffmpeg -i yJbXdOdTaJc.mp4 -ss 23:48 -to 23:53 -crf 23 -r 16 -an clips.mp4

**Comment by daniel-j-h** [comment]
Merging this into master as is right now. We can explore more advanced techniques such as dreaming at multiple scales and related in separate pull requests in the future.
